### ISSUE TITLE
Can `active_dims` really be a slice?

### ISSUE THREAD

**Issue by markvdw** [GPflow/GPflow#223]
@alexisboukouvalas and I were going over the kernexp code today, and we saw that according to the comment in `Kern.__init__()`, `active_dims` it should be possible to pass a `slice` object. Though a `slice` cannot be turned into a numpy array, right? Is this old, or should this be supported?

The reason we're interested in this is because for the `kernexp` branch, we need to check if two kernels work on overlapping dimensions (if so, the kernel expectation contains some extra cross terms on top of the sum of the two kernel expectations). Doing this check is complicated if `active_dims` can also be a non-full slice.

**Comment by jameshensman** [comment]
Yes, I think active dims should be able to be a slice. 

The reason is that a lot of the time we want all of the columns of X, or at least the first D columns of X. Allowing active_dims to be a slice instead of range(...) means a slightly less complicated graph. 

When you're checking for overalapping dimensions, you can interpret a slice s as range(s.start, s.stop, s.step).

**Comment by markvdw** [comment]
Right, but in the current code it doesn't look like a slice is accepted: https://github.com/GPflow/GPflow/blob/master/GPflow/kernels.py#L45

We can't put a slice object in an `np.int32` array!

**Comment by jameshensman** [comment]
Bug. Should accept the slice as self dot active dims.

**Comment by markvdw** [comment]
In order to check for overlapping dimensions, we we need to know `s.stop` though. Any ideas on how to get this information?

**Comment by jameshensman** [comment]
```python
s = slice(10)
print(s.stop)
```

Or did I not understand correctly?

**Comment by markvdw** [comment]
I'm more thinking about this situation:
```
s = np.s_[10:]
print(s.stop)  # None
```
In this case the stop index isn't known by the slice and depends on the size of the matrix that is indexed.

**Comment by jameshensman** [comment]
Oh I see.

I guess if we accept a slice we should make sure it matches `input_dim`, and so must have a stop. Would that work?

**Comment by markvdw** [comment]
So this was something I was confused about as well. Is `input_dim` supposed to be `X.shape[1]` or should it be the number of active dimensions? Since in e.g. `Stationary`, the number of lengthscales is set to `input_dim`. This implies that `input_dim` ought to be the number of active dimensions sliced out of the input, rather than `X.shape[1]`. This also seems to be the case in [1].

This brings me to the conclusion that the kernel does not necessarily have access to `X.shape[1]` at the moment, which would be necessary.

[1] https://github.com/GPflow/GPflow/blob/master/testing/test_kerns.py#L223

**Comment by alexisboukouvalas** [comment]
@markvdw @jameshensman one idea to move this forward for the kernexp/GPLVM branch:

Fix the slice case in kernels to allow one to pass in an arbitrary slice. When checking for non-overlapping active dimensions for the kernel expectations and a slice is present in one of the kernels under Add/Prod, the code can fail gracefully by taking the conservative approach and assuming the dimensions are overlapping. The kernel expectation code will then do quadrature. We may want to emit a warning when doing quadrature.

**Comment by markvdw** [comment]
Fixed with #252, and is addressed in #195.

**Comment by markvdw** [comment]
@jameshensman: So is [kernels.py#L451](https://github.com/GPflow/GPflow/blob/master/GPflow/kernels.py#L451) correct then? If the sub-kernels work on independent dimensions, both will have an `input_dim` of 1, while `Add` should have an `input_dim` of 2. Currently, due to the max being taken, `Add` will have an `input_dim` of 1!

**Comment by jameshensman** [comment]
Correct! We should fix that. If only there were some code to work out whether kernels were overlapping...

**Comment by markvdw** [comment]
The fix I've gone for so far is to require an `input_dim` parameter in `Add`. We could assert that no sub-kernel asks for slices higher than this?

**Comment by markvdw** [comment]
A fix is in #258.

**Comment by jameshensman** [comment]
fixed in #258, closing.
